libflux/src/semantic/nodes.rs

```diff
@@ -573,8 +603,72 @@ pub struct BinaryExpr {
+        // Compute the left and right constraints.
+        // Do this first so that we can return an error if one occurs.
+        let (env, lcons) = self.left.infer(env, f)?;
+        let (env, rcons) = self.right.infer(env, f)?;
+
+        let mut cons = Vec::new();
+        if let Some(kind) = match self.operator {
+            ast::Operator::AdditionOperator => Some(Addable),
+            ast::Operator::SubtractionOperator => Some(Subtractable),
+            ast::Operator::MultiplicationOperator
+            | ast::Operator::DivisionOperator => Some(Divisible),
+            _ => None,
+        } {
+            cons.push(Constraint::Kind(self.typ.clone(), kind));
+        }
+
+        cons.append(&mut Vec::from(match self.operator {
+            // The following operators require both sides to be equal.
+            ast::Operator::AdditionOperator
+            | ast::Operator::SubtractionOperator
+            | ast::Operator::MultiplicationOperator
+            | ast::Operator::DivisionOperator
+            | ast::Operator::PowerOperator
+            | ast::Operator::ModuloOperator => Constraints::from(vec![
+                Constraint::Equal(self.left.type_of().clone(), self.typ.clone()),
+                Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
+            ]),
+            // The following require the type to be a boolean.
+            ast::Operator::GreaterThanEqualOperator
+            | ast::Operator::LessThanEqualOperator
+            | ast::Operator::GreaterThanOperator
+            | ast::Operator::LessThanOperator
+            | ast::Operator::NotEqualOperator
+            | ast::Operator::EqualOperator => {
+                let kind = match self.operator {
+                    ast::Operator::EqualOperator
+                    | ast::Operator::NotEqualOperator => Equatable,
+                    _ => Comparable,
+                };
+                Constraints::from(vec![
+                    Constraint::Kind(self.left.type_of().clone(), kind),
+                    Constraint::Kind(self.right.type_of().clone(), kind),
+                    Constraint::Equal(self.typ.clone(), MonoType::Bool),
+                ])
+            }
+            // Regular expression operators.
+            ast::Operator::RegexpMatchOperator | ast::Operator::NotRegexpMatchOperator => {
+                Constraints::from(vec![
+                    Constraint::Equal(self.left.type_of().clone(), MonoType::String),
+                    Constraint::Equal(self.right.type_of().clone(), MonoType::Regexp),
+                    Constraint::Equal(self.typ.clone(), MonoType::Bool),
+                ])
+            }
+            _ => Constraints::empty(),
+        }));
+
+        // If no constraints were added, then this is an
+        // unsupported binary operator.
+        if cons.is_empty() {
+            return Err(Error::unsupported_binary_operator(&self.operator));
+        }
```

Comment:
I would like to see this error thrown on invalid match arms in the expression above.

Reply:
I'll refactor the match a bit to bring this into the match instead. This is meant to handle the case when `Constraints::empty()` gets returned above, but it can be made more explicit.